File: source/net/auoeke/cin/lexer/Lexer.java
package net.auoeke.cin.lexer;

import java.text.CharacterIterator;
import java.text.StringCharacterIterator;
import java.util.ArrayList;
import java.util.List;
import java.util.ListIterator;
import java.util.function.Consumer;
import java.util.function.Predicate;
import java.util.stream.IntStream;
import net.auoeke.cin.lexer.error.ErrorKey;
import net.auoeke.cin.lexer.error.SyntaxError;
import net.auoeke.cin.lexer.error.SyntaxException;
import net.auoeke.cin.lexer.lexeme.BooleanLexeme;
import net.auoeke.cin.lexer.lexeme.CommaLexeme;
import net.auoeke.cin.lexer.lexeme.CommentLexeme;
import net.auoeke.cin.lexer.lexeme.DelimiterLexeme;
import net.auoeke.cin.lexer.lexeme.FloatLexeme;
import net.auoeke.cin.lexer.lexeme.IntegerLexeme;
import net.auoeke.cin.lexer.lexeme.Lexeme;
import net.auoeke.cin.lexer.lexeme.Lexeme.Type;
import net.auoeke.cin.lexer.lexeme.MappingLexeme;
import net.auoeke.cin.lexer.lexeme.NewlineLexeme;
import net.auoeke.cin.lexer.lexeme.NullLexeme;
import net.auoeke.cin.lexer.lexeme.StringLexeme;
import net.auoeke.cin.lexer.lexeme.WhitespaceLexeme;

public class Lexer implements Iterable<Lexeme> {
    private static final String EXPRESSION_TERMINATORS = "\n={}[]/,";

    private final String cin;
    private final boolean retainSource;
    private final boolean throwExceptions;
    private final StringCharacterIterator iterator;
    private final List<Lexeme> lexemes = new ArrayList<>();

    private int line = 1;
    private int column = 0;
    private int savedLine = -1;
    private int savedColumn = -1;
    private char current;
    private Context context;
    private Lexeme lastCode;

    public Lexer(String cin, boolean retainSource, boolean throwExceptions) {
        this.cin = cin;
        this.retainSource = retainSource;
        this.throwExceptions = throwExceptions;
        this.iterator = new StringCharacterIterator(cin);

        while (this.advance()) {
            this.process(this.current);
        }
    }

    public Lexer(String cin) {
        this(cin, false, true);
    }

    @Override public ListIterator<Lexeme> iterator() {
        return this.lexemes.listIterator();
    }

    private String position() {
        return this.savedLine + ":" + this.savedColumn;
    }

    private int index() {
        return this.iterator.getIndex();
    }

    private int nextIndex() {
        return this.iterator.getIndex() + 1;
    }

    private char next() {
        this.current = this.iterator.current();

        if (this.current == '\n') {
            this.line++;
            this.column = 0;
        }

        this.column++;
        this.iterator.setIndex(this.nextIndex());

        return this.current;
    }

    private boolean advance() {
        return this.next() != CharacterIterator.DONE;
    }

    private char previous() {
        var previous = this.iterator.previous();

        if (previous == '\n') {
            this.line--;
        }

        this.column--;

        return previous;
    }

    private char peek(int distance) {
        var index = this.nextIndex() + distance - 1;
        return index >= this.cin.length() ? CharacterIterator.DONE : this.cin.charAt(index);
    }

    private char peek() {
        return this.peek(1);
    }

    private void savePosition() {
        this.savedLine = this.line;
        this.savedColumn = this.column;
    }

    private boolean whitespaceOrComment(char character) {
        if (character == '\n') {
            this.add(new NewlineLexeme(this.savedLine, this.savedColumn));
        } else if (Character.isWhitespace(character)) {
            this.whitespace(character);
        } else {
            return this.comment(character);
        }

        return true;
    }

    private void requireNext(ErrorKey error, Predicate<Character> predicate) {
        while (this.advance()) {
            this.savePosition();

            if (!this.whitespaceOrComment(this.current)) {
                if (predicate.test(this.current)) {
                    this.process(this.current);
                    return;
                }

                break;
            }
        }

        this.error(error);
    }

    private void requireClose(char character, ErrorKey error) {
        var position = this.position();

        while (this.advance()) {
            if (this.current == character) {
                this.structure(this.current);
                return;
            }

            this.process(this.current);
        }

        this.error(position, error);
    }

    private boolean comma(char character) {
        if (character == ',') {
            this.add(new CommaLexeme(this.savedLine, this.savedColumn));
            return true;
        }

        return false;
    }

    private boolean mapping(char character) {
        if (character == '=') {
            this.add(new MappingLexeme(this.savedLine, this.savedColumn));
            return true;
        }

        return false;
    }

    private void scanExpression(char character, boolean key) {
        if (!key && contains("{}[]", character)) {
            this.structure(character);
        } else if (contains("\"'`", character)) {
            this.string(character);
        } else {
            this.rawString(character);
        }
    }

    private void process(char character, boolean key) {
        this.savePosition();
        if (!this.whitespaceOrComment(character) && !this.comma(character) && !this.mapping(character)) {
            this.scanExpression(character, key);
            if (key) {
                this.requireNext(ErrorKey.NO_MAPPING, s -> contains("={[", s));
            } else {
                while (this.advance()) {
                    if (contains(EXPRESSION_TERMINATORS, this.current) && (this.current != '/' || this.peek() == '/')) {
                        this.previous();
                        return;
                    }

                    if (Character.isWhitespace(this.current)) {
                        this.whitespace(this.current);
                    } else {
                        this.error(ErrorKey.NO_SEPARATOR);
                    }
                }
            }
        }
    }

    private void process(char character) {
        this.process(character, false);
    }

    private void add(Lexeme lexeme) {
        if (lexeme != null && (this.retainSource || !lexeme.type().isSourceOnly())) {
            this.lexemes.add(lexeme);

            if (!lexeme.type().isSourceOnly()) {
                this.lastCode = lexeme;
            }
        }
    }

    private boolean ifNext(char target, Runnable action) {
        var result = this.next() == target;

        if (result) {
            action.run();
        } else {
            this.previous();
        }

        return result;
    }

    private boolean ifNext(char target) {
        return this.ifNext(target, () -> {});
    }

    private void error(String position, ErrorKey error) {
        if (this.throwExceptions) {
            throw new SyntaxException(position, error);
        }

        if (this.lastCode != null) {
            this.lastCode.error = new SyntaxError(error.template.formatted(position));
        }
    }

    private void error(ErrorKey error) {
        this.error(this.position(), error);
    }

    private void whitespace(char whitespace) {
        this.add(new WhitespaceLexeme(this.savedLine, this.savedColumn, buildString(builder -> {
            builder.append(whitespace);

            while (this.advance()) {
                if (!Character.isWhitespace(this.current) || this.current == '\n') {
                    this.previous();
                    return;
                }

                builder.append(this.current);
            }
        })));
    }

    private boolean comment(char character) {
        if (character == '/' && this.ifNext('*')) {
            this.add(new CommentLexeme(this.savedLine, this.savedColumn, Type.BLOCK_COMMENT, buildString(builder -> {
                var depth = 1;

                while (this.advance()) {
                    switch (this.current) {
                        case '/' -> {
                            builder.append('/');

                            var next = this.next();
                            builder.append(next)
                            ;
                            if (next == '*') {
                                depth++;
                            }
                        }
                        case '*' -> {
                            var next = this.next();
                            if (next == '/' && --depth == 0) {
                                return;
                            }
                            builder.append('*');
                            builder.append(next);
                        }
                        default -> builder.append(this.current);
                    }
                }
            })));
        } else if (character == '#' && this.ifNext('#')) {
            this.add(new CommentLexeme(this.savedLine, this.savedColumn, Type.LINE_COMMENT, buildString(builder -> {
                while (this.advance()) {
                    if (this.current == '\n') {
                        this.previous();
                        return;
                    }

                    builder.append(this.current);
                }
            })));
        } else {
            return false;
        }
        return true;
    }

    private void structure(char character) {
        this.add(new DelimiterLexeme(this.savedLine, this.savedColumn, character));

        switch (character) {
            case '}' -> {
                if (this.context != Context.MAP) {
                    this.error(ErrorKey.RBRACE_OUTSIDE_MAP);
                }
            }
            case ']' -> {
                if (this.context != Context.ARRAY) {
                    this.error(ErrorKey.RBRACKET_OUTSIDE_MAP);
                }
            }
            default -> {
                var previousContext = this.context;

                switch (character) {
                    case '{' -> {
                        this.context = Context.MAP;
                        this.requireClose('}', ErrorKey.UNCLOSED_MAP);
                    }
                    case '[' -> {
                        this.context = Context.ARRAY;
                        this.requireClose(']', ErrorKey.UNCLOSED_ARRAY);
                    }
                }

                this.context = previousContext;
            }
        }
    }

    private int delimiterLength(char delimiter) {
        var length = 1;

        while (this.advance()) {
            if (this.current != delimiter) {
                this.previous();
                return length;
            }

            length++;
        }

        return length;
    }

    private void string(char delimiter) {
        var length = this.delimiterLength(delimiter);

        if (length == 2) {
            this.add(new StringLexeme(this.savedLine, this.savedColumn, String.valueOf(delimiter), ""));
        } else {
            this.add(new StringLexeme(this.savedLine, this.savedColumn, String.valueOf(delimiter).repeat(length), buildString(builder -> {
                while (this.advance()) {
                    if (this.current == delimiter) {
                        var count = 1;

                        while (count < length && this.peek(count) == delimiter) {
                            count++;
                        }

                        if (count == length) {
                            repeat(count - 1, this::next);
                            return;
                        }

                        builder.append(this.current);
                        repeat(count - 1, () -> builder.append(this.next()));
                    } else {
                        builder.append(this.current);
                    }
                }

                this.error(ErrorKey.UNCLOSED_STRING);
            })));
        }
    }

    private void rawString(char character) {
        var additionalLexemes = new ArrayList<Lexeme>();

        var string = buildString(builder -> {
            builder.append(character);
            var whitespace = -1;

            while (this.advance()) {
                if (EXPRESSION_TERMINATORS.indexOf(this.current) >= 0 && (this.current != '/' || this.peek() == '/')) {
                    if (whitespace != -1) {
                        additionalLexemes.add(new WhitespaceLexeme(this.savedLine, this.savedColumn, this.cin.substring(whitespace, this.index())));
                    }

                    switch (character) {
                        case '=' -> additionalLexemes.add(new MappingLexeme(this.savedLine, this.savedColumn));
                        default -> this.previous();
                    }

                    return;
                } else if (Character.isWhitespace(this.current)) {
                    if (whitespace == -1) {
                        whitespace = this.index();
                    }
                } else {
                    if (whitespace == -1) {
                        builder.append(this.current);
                    } else {
                        IntStream.rangeClosed(whitespace, this.index()).forEach(index -> builder.append(this.cin.charAt(index)));
                        whitespace = -1;
                    }
                }
            }
        });

        this.add(switch (string) {
            case "false" -> new BooleanLexeme(this.savedLine, this.savedColumn, false);
            case "true" -> new BooleanLexeme(this.savedLine, this.savedColumn, true);
            case "null" -> new NullLexeme(this.savedLine, this.savedColumn);
            default -> {
                try {
                    yield new IntegerLexeme(this.savedLine, this.savedColumn, Long.parseLong(string));
                } catch (NumberFormatException exception) {
                    try {
                        yield new FloatLexeme(this.savedLine, this.savedColumn, Double.parseDouble(string), string);
                    } catch (NumberFormatException e) {
                        yield new StringLexeme(this.savedLine, this.savedColumn, null, string);
                    }
                }
            }
        });

        additionalLexemes.forEach(this::add);
    }

    private static String buildString(Consumer<StringBuilder> builder) {
        var b = new StringBuilder();
        builder.accept(b);
        return b.toString();
    }

    private static void repeat(int n, Runnable action) {
        IntStream.range(0, n - 1).forEach(__ -> action.run());
    }

    private static boolean contains(String string, char character) {
        return string.indexOf(character) >= 0;
    }
}

File: src/net/auoeke/cin/lexer/Lexer.kt
package net.auoeke.cin.lexer

import net.auoeke.cin.lexer.lexeme.*
import net.auoeke.cin.lexer.lexeme.Lexeme.Type
import net.auoeke.extensions.listIterator
import net.auoeke.extensions.repeat
import net.auoeke.extensions.string
import kotlin.collections.ArrayList

@Suppress("UNCHECKED_CAST", "NOTHING_TO_INLINE")
class Lexer(private val cin: String, retainSource: Boolean = false, private val jsonCompatible: Boolean = false, private val throwErrors: Boolean = true) : Iterable<Lexeme> {
    private val iterator: ListIterator<Char> = cin.listIterator()
    private val lexemes = ArrayList<Lexeme>()

    private var line = 1
    private var column = 0
    private var context: Context? = null
    private var elementLexeme: Lexeme? = null

    private inline val position get() = "$line:$column"
    private inline val previousIndex: Int get() = iterator.previousIndex()
    private inline val nextIndex: Int get() = iterator.nextIndex()

    private val acceptsColon get() = jsonCompatible && elementLexeme.let {it is StringLexeme && it.delimiter == "\""}

    init {
        eachChar {
            when {
                else -> process(it)
            }
        }

        eachChar(::process)

        if (!retainSource) {
            lexemes.removeIf {it.type.isComment || it.type == Type.WHITESPACE}
        }
    }

    override fun iterator(): ListIterator<Lexeme> = lexemes.listIterator()

    private fun next() = iterator.next().also {
        if (it == '\n') {
            line++
            column = 0
        }

        column++
    }

    private fun previous() = iterator.previous().also {
        if (it == '\n') {
            line--
        }

        column--
    }

    private fun peek(distance: Int = 0) = cin[nextIndex + distance]

    /** @return `false` if processed. */
    private fun processWhitespaceOrComment(char: Char): Boolean = false.also {
        when {
            char == '\n' -> this += NewlineLexeme(line, column)
            char.isWhitespace() -> whitespace(char)
            char == '/' && ifNext('/') -> comment()
            else -> return true
        }
    }

    private inline fun requireNext(message: String, matchNewline: Boolean = false, predicate: (Char) -> Boolean) = run {
        eachChar {
            when {
                !matchNewline && it == '\n' -> this += NewlineLexeme(line, column)
                it.isWhitespace() -> whitespace(it)
                it == '/' && ifNext('/') -> comment()
                predicate(it) -> {
                    process(it)
                    return
                }
                else -> return@run
            }
        }
    }.let {error(message)}

    private fun require(char: Char, message: String) = position.let {position ->
        eachChar {
            when (it) {
                char -> {
                    structure(it)
                    return
                }
                else -> process(it)
            }
        }

        error(position, message)
    }

    private fun process(char: Char, key: Boolean = false) {
        if (processWhitespaceOrComment(char)) when {
            char == ',' -> comma()
            char == '=' -> mapping(Type.EQUALS)
            char == ':' && acceptsColon -> mapping(Type.COLON)
            else -> {
                when {
                    !key && char in "{}[]" -> structure(char)
                    char in "\"'`" -> string(char)
                    else -> rawString(char, key)
                }

                when {
                    key -> requireNext("%s: Every key in a map must be followed by an assignment, an array or a map.") {
                        it in "={[" || it == ':' && acceptsColon
                    }
                    else -> eachChar {
                        when {
                            key && it == '=' || it in "\n,/}]" -> if (it != '/' || peek() == '/') {
                                previous()
                                return
                            }
                            it.isWhitespace() -> whitespace(it)
                            else -> error("Expected a separator before %s.")
                        }
                    }
                }
            }
        }
    }

    private inline fun eachChar(action: (Char) -> Unit) {
        while (iterator.hasNext()) {
            action(next())
        }
    }

    private infix fun add(lexeme: Lexeme?) {
        lexeme?.let {
            lexemes.add(lexeme)

            when (lexeme.type) {
                Type.COMMA, Type.LINE_COMMENT, Type.BLOCK_COMMENT, Type.NEWLINE, Type.WHITESPACE -> {}
                else -> elementLexeme = lexeme
            }
        }
    }

    private inline operator fun plusAssign(lexeme: Lexeme?) = this add lexeme

    private inline fun ifNext(target: Char, action: () -> Unit = {}): Boolean = (next() == target).also {
        when {
            it -> action()
            else -> previous()
        }
    }

    private fun error(position: String, message: String) = when {
        throwErrors -> throw SyntaxException(position, message)
        else -> elementLexeme?.error = SyntaxError(message)
    }

    private fun error(message: String) = error(position, message)

    private fun whitespace(whitespace: Char) = this add WhitespaceLexeme(line, column, buildString {
        append(whitespace)

        eachChar {
            when {
                !it.isWhitespace() || it == '\n' -> {
                    previous()
                    return@buildString
                }
                else -> append(it)
            }
        }
    })

    private fun comma() = this add CommaLexeme(line, column)

    private fun mapping(operator: Type) = this add MappingLexeme(line, column, operator)

    private fun comment() = this add when {
        ifNext('/') -> CommentLexeme(line, column, Type.BLOCK_COMMENT, buildString {
            eachChar {
                if (it == '/' && next() == '/') {
                    when (val char = next()) {
                        '/' -> return@buildString
                        else -> append(char)
                    }

                    append('/')
                }

                append(it)
            }
        })
        else -> CommentLexeme(line, column, Type.LINE_COMMENT, buildString {
            eachChar {
                if (it == '\n') {
                    previous()
                    return@buildString
                }

                append(it)
            }
        })
    }

    private fun structure(char: Char) {
        this += DelimiterLexeme(line, column, char)

        when (char) {
            '}' -> if (context !== Context.MAP) {
                error("Found '$char' at %s outside a map.")
            }
            ']' -> if (context !== Context.ARRAY) {
                error("Found '$char' at %s outside an array.")
            }
            else -> context = context.also {
                when (char) {
                    '{' -> {
                        context = Context.MAP
                        require('}', "%s: Map is not closed.")
                    }
                    '[' -> {
                        context = Context.ARRAY
                        require(']', "%s: Array is not closed.")
                    }
                }
            }
        }
    }

    private fun delimiterLength(delimiter: Char): Int {
        var length = 1

        eachChar {
            if (it == delimiter) {
                length++
            } else {
                previous()
                return length
            }
        }

        return length
    }

    private fun string(delimiter: Char) = this add when (val length = delimiterLength(delimiter)) {
        2 -> StringLexeme(line, column, delimiter.string, "")
        else -> StringLexeme(line, column, delimiter.repeat(length), buildString {
            eachChar {
                if (it != delimiter) {
                    append(it)
                } else {
                    val next = next()

                    if (next != delimiter) {
                        append(it)
                        append(next)
                    } else {
                        val next1 = next()

                        if (next1 == delimiter) {
                            return@buildString
                        }

                        append(it)
                        append(next)
                        append(next1)
                    }
                }
            }

            error("%s: The string is not closed.")
        })
    }

    private fun rawString(char: Char, key: Boolean = false) {
        val additionalLexemes = ArrayList<Lexeme>()

        this += when (val string = buildString {
            append(char)

            var whitespace: Int = -1

            eachChar {
                when {
                    key && char in "={[" || !key && it in "\n,/}]" -> if (it != '/' || peek() == '/') {
                        if (whitespace != -1) {
                            additionalLexemes += WhitespaceLexeme(line, column, cin.substring(whitespace, previousIndex))
                        }

                        when (char) {
                            '=' -> additionalLexemes += MappingLexeme(line, column, Type.EQUALS)
                            else -> previous()
                        }

                        return@buildString
                    }
                    else -> when {
                        it.isWhitespace() -> if (whitespace == -1) {
                            whitespace = previousIndex
                        }
                        else -> when (whitespace) {
                            -1 -> append(it)
                            else -> {
                                whitespace.rangeTo(previousIndex).forEach {index -> append(cin[index])}
                                whitespace = -1
                            }
                        }
                    }
                }
            }
        }) {
            "false" -> BooleanLexeme(line, column, false)
            "true" -> BooleanLexeme(line, column, true)
            "null" -> NullLexeme(line, column)
            else -> try {
                IntegerLexeme(line, column, string.toLong())
            } catch (exception: NumberFormatException) {
                try {
                    FloatLexeme(line, column, string.toDouble(), string.endsWith('.'))
                } catch (exception: NumberFormatException) {
                    StringLexeme(line, column, null, string)
                }
            }
        }

        additionalLexemes.forEach(::add)
    }
}
